sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+following information:
+
+* The source position of the `[[sycl::requires()]]` attribute that decorates
+  the kernel or device function which provides the `Allowed` aspect set.  This
+  tells the user which aspects the kernel or device function intends to use.
+
+* The source position of the call to a function that is decorated with
+  `[[sycl::requires()]]` or the source position of the use of a type that is
+  decorated with `[[sycl::requires()]]`.  This tells the user where in the
+  kernel a particular aspect is actually used.
+
+Note that this analysis can be done in the front-end compiler even when a
+kernel makes a call to a function that is in another translation unit.
+Language rules require the application to declare such a function with
+`SYCL_EXTERNAL` in the calling TU, and the `SYCL_EXTERNAL` declaration must be
+decorated with the `[[sycl::requires()]]` attribute.  Therefore, the front-end
+can diagnose errors with aspect usage even without seeing the definition of the
+`SYCL_EXTERNAL` function.
+
+
+## Design to raise required exceptions (and avoid forbidden errors)
+
+As described above the runtime must raise an `errc::kernel_not_supported`
+exception when a kernel is submitted to a device that does not support the
+optional features that the kernel uses.  Likewise, the runtime must **not**
+raise an exception (or otherwise produce an error) due to speculative
+compilation of a kernel for a device, unless the application actually submits
+the kernel to that device.  The solution is largely the same for both JIT and
+AOT cases.
+
+### JIT case
+
+The JIT case requires some change to the way kernels are bundled together into
+device images.  Currently, kernels are bundled together regardless of the
+features they use, and this can lead to JIT errors due to speculative
+compilation.  Consider a device image that contains two kernels: `K1` uses no
+optional features and `K2` uses an optional feature that corresponds to aspect
+`A`.  Now consider that the application submits kernel `K1` to a device that
+does not support aspect `A`.  Since the two kernels are bundled together into
+one device image, the runtime really compiles both kernels for the device.
+Currently, this will raise a JIT exception because the compilation of kernel
+`K2` will fail when compiled for a device that does not support aspect `A`.
+
+There are two ways to solve this problem.  One is to change the way kernels are
+bundled into device images such that we never bundled two kernels together
+unless they required exactly the same set of device aspects.  Doing this would
+avoid the error described above.  However, we have elected for a different
+solution.
+
+Instead, we will allow kernels to be bundled together as they currently are,
+but we will introduce extra decorations into the generated SPIR-V that allow
+the JIT compiler to discard kernels which require aspects that the device does
```

Comment:
+100 for requiring BE (JIT) to honor required aspects! That's the right way, I believe.